### PR TITLE
sdn: OVS pods require more memory

### PR DESCRIPTION
**Pull request by dcbw**
Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1571379

@jeremyeder @smarterclayton @knobunc

### CODE REVIEW

**Comment by smarterclayton** [comment]
We should not be setting limits on these - who puts those there?  @derekwaynecarr ?

**Comment by smarterclayton** [comment]
Why is openshift-sdn using so much memory?  Caches?  Is the amount enough to ensure all possible clusters in the world will be able to run it?

File: roles/openshift_sdn/files/sdn-ovs.yaml [hunk 1]
```diff
@@ -83,10 +83,10 @@ spec:
         resources:
           requests:
             cpu: 100m
-            memory: 200Mi
```
Comment (smarterclayton) [inline roles/openshift_sdn/files/sdn-ovs.yaml]: Why bump request?  What's the floor for a 10 node 100 pod cluster?

**Comment by dcbw** [comment]
> Why is openshift-sdn using so much memory? Caches? Is the amount enough to ensure all possible clusters in the world will be able to run it?

@smarterclayton in this case it's the OVS container, not the SDN one.  That consists of the vswitchd, ovsdb-server, and the monitoring processes that are supposed to restart them if they die abnormally.

File: roles/openshift_sdn/files/sdn-ovs.yaml (hunk 1 above)
Comment (jeremyeder) [reply roles/openshift_sdn/files/sdn-ovs.yaml]: 58MB in my tests

**Comment by jeremyeder** [comment]
Some research into previous tests and prod environments on this topic...
starter
```
free-stg, ocp3.9: avg RSS 58MB
free-int, ocp3.10: avg RSS 60MB
```

previous scale lab tests:
```
ocp3.7 up to 360MB on infra nodes, 170MB elsewhere (ovs-2.7)
ocp3.9 up to 380MB on infra nodes, 180MB elsewhere (ovs-2.7)
```

Now in the lab w/3.10, see how the RSS varies pretty wildly.  There are just ~100 pods in this environment, all our own stuff (logging/prom/cns/servicecatalog/all the new daemonsets).

```
ocp3.10
cns-1.scale-ci.example.com >> 157352
app-node-0.scale-ci.example.com >> 58732
cns-0.scale-ci.example.com >> 157360
app-node-1.scale-ci.example.com >> 58600
cns-2.scale-ci.example.com >> 157344
master-2.scale-ci.example.com >> 157036
master-0.scale-ci.example.com >> 156768
master-1.scale-ci.example.com >> 157448
infra-node-2.scale-ci.example.com >> 353496
infra-node-0.scale-ci.example.com >> 353524
infra-node-1.scale-ci.example.com >> 353632
```

* We will need to do more research into why the differences in RSS memory usage between nodes, as well as why we see more RSS in the lab than in other environments.

* So far in all our scale tests, the high water mark for RSS on infra nodes is below 400Mi.

**Comment by jeremyeder** [comment]
I filed https://bugzilla.redhat.com/show_bug.cgi?id=1572619 to follow up with openvswitch folks on why ovs-vswitchd uses more memory on infra nodes.

**Comment by dcbw** [comment]
> I filed https://bugzilla.redhat.com/show_bug.cgi?id=1572619 to follow up with openvswitch folks on why ovs-vswitchd uses more memory on infra nodes.

@jeremyeder ah missed that you'd filed one already; I also filed one with more specific details of what's happening.  So I duped yours to mine: https://bugzilla.redhat.com/show_bug.cgi?id=1572797

**Comment by dcbw** [comment]
Anyway, short summary is that ovs-vswitchd will create #cores+1 pthreads and RHEL/Fedora default to 8MB of pthread stack.  The vswitch will immediately dirty that stack (this is the "bug" in all this), resulting in #cores+1 * 8MB which is where you're seeing the wildly varying memory usage between nodes as each node class has a different # of vCPUs.  The infra nodes had 40 vCPUs which is why they were blowing the previous 300MB OVS container limits but the app nodes weren't.

Suggested workaround for now is to:

ovs-vsctl set Open_vSwitch . other_config:n-revalidator-threads=X
ovs-vsctl set Open_vSwitch . other_config:n-handler-threads=Y

Where say X=2 and Y=5 (default is ~1:2.5 revalidator:handler), which would result in 64M of RSS devoted to pthreads.  Still better than 320MB that 40 cores would use by default.

**Pull request merged**